### PR TITLE
HPCC-32455 Upgrade WsSQL instrumentation

### PR DESCRIPTION
**Pull request by timothyklemm**
Replace eight pairs of trace summary timestamps with eight internal spans.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_sql/ws_sqlService.cpp [hunk 1]
```diff
@@ -986,10 +1008,17 @@ bool CwssqlEx::onExecuteSQL(IEspContext &context, IEspExecuteSQLRequest &req, IE
...
 
-                context.addTraceSummaryTimeStamp(LogNormal, "strtWUCompile");
-                WsWuHelpers::submitWsWorkunit(context, compiledwuid.str(), cluster, nullptr, 0, 0, true, false, false, nullptr, nullptr, nullptr, nullptr);
-                waitForWorkUnitToCompile(compiledwuid.str(), req.getWait());
-                context.addTraceSummaryTimeStamp(LogNormal, "endWUCompile");
+                OwnedSpanScope wuCompileSpanScope(queryThreadedActiveSpan()->createInternalSpan("submit_ws_workunits"));
+                try
+                {
+                    WsWuHelpers::submitWsWorkunit(context, compiledwuid.str(), cluster, nullptr, 0, 0, true, false, false, nullptr, nullptr, nullptr, nullptr);
+                    waitForWorkUnitToCompile(compiledwuid.str(), req.getWait());
```
Comment (rpastrana) [inline esp/services/ws_sql/ws_sqlService.cpp:1015]: it might help to annotate the span w/ the wuid

File: esp/services/ws_sql/ws_sqlService.cpp [hunk 2]
```diff
@@ -1022,17 +1051,33 @@ bool CwssqlEx::onExecuteSQL(IEspContext &context, IEspExecuteSQLRequest &req, IE
 
             if (clonable)
             {
-                context.addTraceSummaryTimeStamp(LogNormal, "StartWUCloneExe");
-                cloneAndExecuteWU(context, compiledwuid.str(), runningwuid, xmlparams.str(), NULL, NULL, cluster);
-                context.addTraceSummaryTimeStamp(LogNormal, "EndWUCloneExe");
+                OwnedSpanScope wuCloneSpanScope(queryThreadedActiveSpan()->createInternalSpan("clone_and_execute_wu"));
+                try
+                {
+                    cloneAndExecuteWU(context, compiledwuid.str(), runningwuid, xmlparams.str(), NULL, NULL, cluster);
```
Comment (rpastrana) [inline esp/services/ws_sql/ws_sqlService.cpp:1057]: annotate span w/ both wuids, also cluster might be important to add

File: esp/services/ws_sql/ws_sqlService.cpp [hunk 3]
```diff
@@ -1041,9 +1086,16 @@ bool CwssqlEx::onExecuteSQL(IEspContext &context, IEspExecuteSQLRequest &req, IE
             int timeToWait = req.getWait();
             if (timeToWait != 0)
             {
-                context.addTraceSummaryTimeStamp(LogNormal, "StartWUProcessWait");
-                waitForWorkUnitToComplete(runningwuid.str(), timeToWait);
-                context.addTraceSummaryTimeStamp(LogNormal, "EndWUProcessWait");
+                OwnedSpanScope wuProcessSpanScope(queryThreadedActiveSpan()->createInternalSpan("wait_for_workunit_to_complete"));
+                try
+                {
+                    waitForWorkUnitToComplete(runningwuid.str(), timeToWait);
```
Comment (rpastrana) [inline esp/services/ws_sql/ws_sqlService.cpp:1092]: I'm not sure where this function is defined, but seems worth tracing it (perhaps outside the scope of this PR)

**Review by rpastrana — COMMENTED** [review]
@timothyklemm looks good, there's some basic info we could add to the new spans.

File: esp/services/ws_sql/ws_sqlService.cpp (hunk 3 above)
Comment (timothyklemm) [reply esp/services/ws_sql/ws_sqlService.cpp:1092]: @rpastrana the method is defined in workunit.cpp. It redirects through a factory to access one of multiple implementations (one of which is "unimplemented"). The method is called from multiple services which are not instrumented with tracing. Even within this service, only the calls being made within traced methods are included in child spans.

I think a separate PR is appropriate for changing the common function. Changes in this service can be made as part of this PR after additional discussion.

**Review by rpastrana — APPROVED** [review]
@timothyklemm approved

**Comment by timothyklemm** [comment]
@ghalliday please merge.

**Review by ghalliday — APPROVED** [review]
Any reason why this should go in 9.8.x rather than 9.10.x?  If not I will re-target and merge to master.

**Comment by timothyklemm** [comment]
> Any reason why this should go in 9.8.x rather than 9.10.x? If not I will re-target and merge to master.

I know it started on 9.8.x before 9.10.x. I'm not aware of any reason it needs to stay there.

**Pull request merged**